AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/SystemKey/ContentHandler.dfy

```diff
@@ -0,0 +1,261 @@
+      input.Crypto);
+    hide *;
+    var kmsReq := KMS.Types.DecryptRequest(
+      KeyId := Some(input.MaterialIdentifier),
+      CiphertextBlob := input.CiphertextBlob,
+      EncryptionContext := Some(kms_ec),
+      GrantTokens := Some(input.KmsTuple.grantTokens)
+    );
+    var kmsRes? := input.KmsTuple.kmsClient.Decrypt(kmsReq);
+    // var kmsRes? := kmsRes?; //.MapFailure(e => Types.ComAmazonawsKms(e));
+    if (kmsRes?.Failure?) {
+      if (kmsRes?.error.InvalidCiphertextException?) {
+        return Success(false);
+      } else {
+        // TODO-Mutations-GA : better error message for failure
+        // Otherwise, it is some other KMS issue, and we return a KeyStoreAdminException
+        return Failure(Types.ComAmazonawsKms(ComAmazonawsKms:=kmsRes?.error));
+      }
+    }
+    var kmsRes := kmsRes?.value;
+    :- Need(
+      kmsRes.Plaintext.Some?,
+      // TODO-Mutations-GA : better error message for failure
+      Types.KeyStoreAdminException(message := "KMS returned an invalid response.")
+    );
+    // =- 4. Assert plain-text is "Sign"
+    :- Need(
+      kmsRes.Plaintext.value == SIGN_UTF8_BYTES,
+      // TODO-Mutations-GA : better error message for failure
+      Types.KeyStoreAdminException(message := "KMS returned an invalid response.")
+    );
+    return Success(true);
+
+  }
+
+  method {:vcs_split_on_every_assert} ContentToKmsEncryptionContext(
+    Content: Content,
+    Crypto: AtomicPrimitives.AtomicPrimitivesClient
+  )
+    returns (output: Result<KMS.Types.EncryptionContextType, SerializeError>)
+    requires Crypto.ValidState() && Content.ValidState()
+    ensures Crypto.ValidState()
+    modifies Crypto.Modifies
+    ensures output.Failure? ==> output.error.KeyStoreAdminException?
+  {
+    hide *;
+    //=- 1. EncryptionContextDigest
+    var digestResult: Result<seq<uint8>, CanonicalEncryptionContext.CanonizeDigestError> :=
+      CanonicalEncryptionContext.EncryptionContextDigest(Crypto, Content.ContentToSHA);
+    if (digestResult.Failure?) {
+      var error: Types.Error;
+      error := match digestResult.error {
+        case AwsCryptographyPrimitives(e) =>
+          // we cannot reliably serialize a Primitive error without work
+          Types.KeyStoreAdminException(message:="Could not SHA-384 Content.")
+        case AwsCryptographicMaterialProvidersException(e) =>
+          Types.KeyStoreAdminException(message:="Could not SHA-384 Content. " + e)
+      };
+      return Failure(error);
+    }
+    //=- 2. Base64 Encode
+    var encodeResult: Result<MPL.Types.Utf8Bytes, string> :=
+      UTF8.Encode(Base64.Encode(digestResult.value));
+    if (encodeResult.Failure?) {
+      var error := Types.KeyStoreAdminException(
+        message:="Could not serialize Digest of content. " + encodeResult.error
+      );
+      return Failure(error);
+    }
+    //=- 3. Base64 encoded content in EC, along with parition key and sort key
+    // Dafny forgets that DIGEST_UTF8_BYTES is valid
```

Comment:
It's not that Dafny forgets; it is that atm the definition of the constant is just a seq<uint8> (we know that they are utf8 bytes but dafny has not been told they are. by calling the method that properly encodes and proofs that the output is validutfseq dafny has no problem with the proof